### PR TITLE
Add more diff size limits

### LINKED ISSUE #2596: introduce some soft limits around diff rendering
The client apps have historically been indifferent to large diffs, whereas GitHub the Website is very rigid about some diffs - requiring users to opt-in to view them.

screen shot 2017-09-01 at 9 14 53 am

As discussed in https://github.com/desktop/desktop/issues/1686#issuecomment-326443927 - there may be some diff sizes and shapes that make the editor cry (ridiculously long lines with wrapping enabled seems to be what currently hurts Codemirror). Until we have time to revisit our choice of diff experience (post 1.0 at least), I'd love to get some feels from @desktop/core on warning users about large diffs to keep things snappy.

For sake of simplicity and not using it as crutch, I'd rather not make this configurable at all to the user - for diffs over a certain size, display a message first and require the user to click to see it.

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #2596

This introduces two more limits:

1. A diff cannot be over 3MB. This is to catch any generally big diff.
1. A diff cannot have a line longer than 500,000 characters, e.g., minified files.

Both of those are pretty arbitrary, determined by testing what felt slow on my computer 😬 We can tweak them based on feedback.

I created test files with `base64 /dev/urandom | head -c 10000000 > file.txt`.

I _didn't_ implement the ability to show the diffs anyways. Given that that experience means the app freezes, I'm not sure we wanna allow that without taking a harder look at diff rendering performance.

### CODE REVIEW

File: app/src/lib/git/diff.ts [hunk 1]
```diff
@@ -25,13 +24,56 @@ import { spawnAndComplete } from './spawn'
...
+  return buffer.length < MaxDiffBufferSize
+}
+
+/** Is the buffer too large for us to reasonably represent? */
+function isBufferTooLarge(buffer: Buffer) {
+  if (!isValidBuffer(buffer)) {
+    return true
+  }
+
+  return buffer.length >= MaxReasonableDiffSize
```
Comment (niik) [inline app/src/lib/git/diff.ts]: Could this be `isValidBuffer(buffer) && buffer.length < MaxReasonableDiffSize`?

**Comment by joshaber** [comment]
🐸

File: app/src/lib/git/diff.ts (hunk 1 above)
Comment (niik) [reply app/src/lib/git/diff.ts]: lol, I got those confused. I see you understood what I meant tho ✨

Comment (joshaber) [reply app/src/lib/git/diff.ts]: Throw a De Morgan's Law on it 💥

**Comment by jessetan** [comment]
@joshaber I'm still encountering slow diffs on some files. The max line length of 500,000 seems pretty high (mine has lines 111,000 chars long and takes 15s to diff and syntax highlight). I would argue that any file that has lines of over a few hundred chars is not handwritten, but probably minified. Perhaps a limit of 1000-10,000 chars per line would be better?

**Comment by joshaber** [comment]
Thanks for the feedback @jessetan! I agree we could cut it down even more. I opened https://github.com/desktop/desktop/issues/3740 to track that.

**Pull request merged**